File: 06-hash-tables-maps-sets/09-two-sum/two-sum.js
function twoSum(nums, target) {
  const newSet = new Set();

  for (let i = 0; i < nums.length; i++) {
    const compliment = target - nums[i];

    if (newSet.has(compliment)) {
      return [nums.indexOf(compliment), i];
    }

    newSet.add(nums[i]);
  }

  return [];
}

console.log(twoSum([2, 7, 11, 15], 9));

module.exports = twoSum;

File: 06-hash-tables-maps-sets/09-two-sum/TwoSum.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int compliment = target - nums[i];
            
            if (numMap.containsKey(compliment)) {
                return new int[]{numMap.get(compliment), i};
            }
            
            numMap.put(nums[i], i); // Store the number and its index
        }
        
        return new int[]{}; // Return empty array if no solution is found
    }

    public static void main(String[] args) {
        int[] result = twoSum(new int[]{2, 7, 11, 15}, 9);
        System.out.println("[" + result[0] + ", " + result[1] + "]");
    }
}
